partialmessages/partialmsgs.go

```diff
@@ -0,0 +1,290 @@
+package partialmessages
+
+import (
+	"bytes"
+	"errors"
+	"iter"
+	"log/slog"
+	"slices"
+
+	pb "github.com/libp2p/go-libp2p-pubsub/pb"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// TODO: Add gossip fallback (pick random connected peers and send ihave/iwants)
+// TODO: Update scoring code to not kick partial-message only peers
+// Question: How to configure scheduling RPCs?
+//   - Leave it up to the implementation of sendRPC
+// Question: Skip partial IHAVE for now?
+// Question: I could have a user provided validation queue instead of requiring republishing
+//   - But a user may need to republish anyways if they get parts out of band
+
+const minGroupTTL = 3
+
+// PartialMessage is a message that can be broken up into parts. It can be
+// complete, partially complete, or empty. It is up to the application to define
+// how a message is split into parts and recombined, as well as how missing and
+// available parts are represented.
+//
+// It is passed to Gossipsub with a PublishPartialMessage method call. Gossipsub
+// keeps a reference to this object, so the implementation should either not
+// mutate this object in a separate goroutine after handing it to Gossipsub, or
+// take care to make the object thread safe.
+type PartialMessage interface {
+	GroupID() []byte
+
+	// PartialMessageBytes takes in the opaque request metadata and
+	// returns a encoded partial message that fulfills as much of the request as
+	// possible. It also returns a opaque request metadata representing the
+	// parts it could not fulfill. This MUST be empty if the implementation could
+	// fulfill the whole request.
+	//
+	// An empty metadata should be treated the same as a request for all parts.
+	//
+	// If the Partial Message is empty, the implementation MUST return:
+	// nil, metadata, nil.
+	PartialMessageBytes(metadata []byte) ([]byte, []byte, error)
+
+	// PartsMetadata returns metadata about the parts this partial message
+	// contains and, possibly implicitly, the parts it wants.
+	PartsMetadata() []byte
+}
+
+// partialMessagePeerState is the state we keep per peer. Used to make Publish
+// Idempotent.
+type partialMessagePeerState struct {
```

Comment:
NIT: for the same reasons as the renaming for PartialMessage. 
```suggestion
type peerState struct {
```